### ISSUE TITLE
Is there a rule to detect bad exit() calls?

### ISSUE THREAD

**Issue by afbarbaro** [vimeo/psalm#3425]
We recently run into an issue where a method had a `exit();` call inside it that was inadvertently not caught in code review (😞 😬).  I was wondering if `Psalm` can detect this and so our build would have failed preventing us from introducing this issue in our code.

Thank you

**Comment by afbarbaro** [comment]
https://psalm.dev/r/7d116e039a

**Comment by weirdan** [comment]
There's `<forbiddenFunctions>` config element, but, unfortunately, `exit()` is not a function.

**Comment by weirdan** [comment]
From the type system perspective, Psalm might require (rather than just allowing) adding `no-return` type to `@return` docblock for any potentially terminating function. If this was implemented, you'd still be able to use `exit`, but would have to confirm your intention like this:
```php
/** @return int|no-return **/
function mayExit(): int {
   if (rand(0, 1)) {
        exit;
   }
   return 42;
}
```
This would also apply to any callers, callers of the callers, etc. Which probably means too much noise for too many projects to be practical.

**Comment by weirdan** [comment]
Could be implemented similar to https://github.com/vimeo/psalm/blob/6784a90b2ff6a93ad57cd093ef2ccfbe65e1da54/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php#L351-L363

@afbarbaro mind doing the PR?

**Comment by afbarbaro** [comment]
Awesome! Thanks.  I'll take a stab at it!

**Comment by bdsl** [comment]
@weirdan @afbarbaro  I'm not sure this makes sense for the type system - as I understand it `no-return` is the [bottom type](https://en.wikipedia.org/wiki/Bottom_type), which is a subtype of every other type. The union of `int` and `no-return` is `int`.

Exit isn't the most usual way for a function to not return - throwing an exception is much more frequent. I certainly wouldn't want to have to change the type declaration for any function that can throw.

**Comment by orklah** [comment]
This was implemented a week ago
